### DISCUSSION TITLE
Failure to upgrade server's hardware and re-install Security Onion on new server

### DISCUSSION THREAD

**Discussion by McLoggin** [Security-Onion-Solutions/securityonion#5786] [Unsupported Versions] [answered]
We're doing a Cent7 network install on version 2.3.80 for our distributed deployment, and we swapped out servers recently to upgrade the hardware on one of our sensors. I followed the docs recommendation for removing a node from the cluster, https://docs.securityonion.net/en/latest/removing-a-node.html Now I'm running into an issue where we are installing it on a node that has the same hostname and IP address as the previous server that was working just fine. When I check the sensor's sls file in /opt/so/saltstack/local/salt/minions/[hostname]_sensor.sls, it shows the right interfaces, hostname, and IPs. I've tried multiple times to re-install the OS multiple times, and re-download security onion to install it on the sensor. This is the error from /root/sosetup.log from the sensor:

![SosetupErrors](https://user-images.githubusercontent.com/80071690/136098923-e077181b-08ea-4478-9268-1bd15d214978.png)

I've tried to follow the error's recommendations and delete the key again, and restart the sensor. It still fails with the same error.

When I run salt-key -L after trying the install that fails on the new server, it shows that it has accepted the salt-key from that sensor. Although the install fails and states that it rejected the salt-key. If I try to re-run so-setup-network without deleting the key, it fails on the same error.

Here is the error from the management server in /opt/so/log/salt/master.log. Is there something that I'm missing to fully remove the node, so I can add it back to the cluster?

![masterLogError](https://user-images.githubusercontent.com/80071690/136099829-55f3ca80-229f-4499-8329-8cc6dec83123.png)

Is there someway that I could remove any reference to that minionID, I'm wondering if that could be causing the issue?

**Comment by dougburks** [comment, marked as answer]
So just to confirm, did you follow all the steps at https://docs.securityonion.net/en/latest/removing-a-node.html including both of the following?

- sudo salt-key -d nodename

AND

- Remove the node from any .sls files in /opt/so/saltstack/local/pillar/data/.

Additionally, have you tried removing the file for that minion from `/opt/so/saltstack/local/pillar/minions/` and then running Setup on the minion to see if it will write a new minion file successfully?

**Reply by McLoggin** [reply]
Yes sir, I tried removing the sls file from /opt/so/saltstack/local/pillar/minions, along with removing the salt-key and the entry from /opt/so/saltstack/local/pillar/data/sensorstab.sls. Also, I delete any old public keys from that server in the authorized_keys file for soremote user. Another thing I did was delete any entry for that IP in /opt/so/saltstack/local/salt/firewall/*. Then, I rebooted the management server. Then, I went ahead and tried re-installing on the sensor.
 
Upon re-installing Cent7 and running so-setup-network again, it does create a file in /opt/so/saltstack/local/pillar/minions. Also, it creates an entry in /opt/so/saltstack/local/pillar/data/sensorstab.sls, and it transfers the host's public key to the authorized_keys file for soremote. The entries are put back into /opt/so/saltstack/local/salt/firewall. One interesting thing I'm seeing is that I can run test.ping using salt from the management node against that sensor. Also, I noticed that there is no entry for the new sensor in Grid within the SOC dashboard.

![test ping](https://user-images.githubusercontent.com/80071690/136392583-08d7250d-fd9e-4334-bafd-7c8da7aeae5d.png)

When I try to run state.highstate from the management node against that sensor, it fails almost immediately with the error below about the key. Although when I check to see if the management node has accepted the key using salt-key -L, it shows that it has accepted that key from that sensor.


![state highstate](https://user-images.githubusercontent.com/80071690/136393243-17a81477-9b13-45ab-88e1-fc33fc9a0219.png)

On the sensor's side of things, it creates a directory for /opt/containerd and /opt/so/log/salt/minion. The directories /opt/so/saltstack/default & /opt/so/saltstack/local are also created but nothing is in those directories. The only entries in /opt/so/log/salt/minion is the exact same as the first screenshot from my original post above. None of the so-<service> commands are installed on the sensor though. Docker has been installed on the sensor, and is actively running. None of the docker images get pushed to the sensor, confirmed with docker image ls. Docker ps doesn't return any results as well. Checking /root/so-setup.log, shows this error. Within the gui after running ./so-setup-network, it does prompt me to accept the host key and enter the password, which I did like any other SO2 install.


![sosetup log](https://user-images.githubusercontent.com/80071690/136395401-18689c14-82bb-4dbe-b311-727a6b861c23.png)

On a separate note, just because I've seen this asked in other discussion post. It can't be anything to do with a firewall because it was working just fine before, and I don't see any changes made on that network. All the other nodes in the cluster are working, I have a couple more servers that I need to upgrade. I'm holding out on doing them until I can try to get this one fixed though. The server upgrades are pretty beefy, so I know that the hardware isn't holding anything back.


Appreciate your help!

**Reply by dougburks** [reply]
Is your manager running 2.3.80?

After removing the old sensor's information from the manager, have you tried restarting the salt services on the manager or rebooting it altogether to eliminate any possible issues with caching?

**Reply by McLoggin** [reply]
Yessir, our whole fleet is on 2.3.80. So, the last install that I tried. I did all that I said in the post above, then an hour later I rebooted the manager entirely. Finally thats when I re-installed the OS onto the sensor and tried the network install. It couldn't have been a caching problem because it wasn't calling that sensor for salt anymore. Something I did notice was that the entry was still in the Grid dashboard, but that when away after I rebooted the manager.

**Reply by dougburks** [reply]
You might want to set up a test deployment in VMs to see if you can duplicate this issue. If so, VM snapshot capability may make it easier and faster to troubleshoot.

**Reply by McLoggin** [reply]
I figured it out, a single quote inside of /opt/so/saltstack/local/pillar/minions/<sensor_name>_sensor.sls within the node_description: entry caused the installation to fail.

When going through the installation and it came to Node's Description, I put "<SensorName>'s sensor", this single quote inside of the Node Description gets parsed like node_description: 'Sierra's Sensor', and it fails on that because it ends the node's description after the second single quote. Causing the error because the rest of the description wasn't parsed correctly after 's sensor' in the Node Description. Which doesn't follow standard yaml syntax

**Reply by dougburks** [reply]
I've created Issue https://github.com/Security-Onion-Solutions/securityonion/issues/5937 to try to prevent this in the future.
